### PR TITLE
[FABN-835] Refactor CryptoKeyStore (Part 3)

### PR DESCRIPTION
**Pull request by davidkhala**
- remove constructor function in Utils.js
- move newInstance function logic to Utils.js, making CryptoKeyStore exportable

Signed-off-by: david <david-khala@hotmail.com>

### CODE REVIEW

File: fabric-common/lib/Utils.js
Comment (bestbeforetoday) [inline fabric-common/lib/Utils.js:467]: Does this mean that `cks.initialize()` might be called more than once if another call of `cks._getKeyStore()` passes this line before a previous call completes the await on `cks.initialize()` on the following line and sets `cks.isReady` to `true`? I wonder if something like this might be more reliable:
```
const cks = new CryptoKeyStore(kvs);
const isReady = cks.initialize();
cks._getKeyStore = async () => {
    await isReady;
    return cks;
};
```
It does mean `cks.initialize()` gets kicked off before anyone calls `cks._getKeyStore()` though. I don't know if that is a problem.

Comment (bestbeforetoday) [reply fabric-common/lib/Utils.js:467]: Now I think about it, I'm not really sure what the point is of having the `cks._getKeyStore()` at all, since it just calls initialize() and returns `cks`. It seems to me that it might be simpler to have `newCryptoKeyStore` as an async factory function something like:
```
module.exports.newCryptoKeyStore = async (KVSImplClass, opts) => {
    const kvs = typeof KVSImplClass === 'function' ? new KVSImplClass(opts) : new InMemoryKeyValueStore();
    const CryptoKeyStore = require('./impl/CryptoKeyStore');
    const cks = new CryptoKeyStore(kvs);
    await cks.initialize();
    return cks;
};
```
Maybe it's just to avoid changing `newCryptoKeyStore` to an async function and requiring calling code changes?

Even better might be to push this async factory function into the CryptoKeyStore module and export that instead of the CryptoKeyStore class itself so calling code can't get it wrong and fail to initialize a new CryptoKeyStore.

Comment (davidkhala) [reply fabric-common/lib/Utils.js:467]: Reply to “Maybe it's just to avoid changing newCryptoKeyStore to an async function and requiring calling code changes?”
Yes.

But for the last section: 
"Even better might be to push this async factory function into the CryptoKeyStore module and export that instead of the CryptoKeyStore class itself so calling code can't get it wrong and fail to initialize a new CryptoKeyStore."

I am not sure I get the idea. Do you mean we adding a new static method inside `CryptoKeyStore` such as 
```
async static newCryptoKeyStore (kvs) {
  
    const cks = new CryptoKeyStore(kvs);
    await cks.initialize();
    return cks;
}
```

Comment (bestbeforetoday) [reply fabric-common/lib/Utils.js:467]: I was thinking it could be a module-scope function. Then only that function could be exported and the actual CryptoKeyStore class could be entirely private within the module (not exported). So the only way to access the CryptoKeyStore would be by calling `newCryptoKeyStore()` and there would be no possible way (without subverting the Node module system) for any calling code to fail to initialize it correctly.

That does assume that no other code needs to directly access the CryptoKeyStore class though, such as to extend it. I am not sure whether or not that is the case.

Comment (davidkhala) [reply fabric-common/lib/Utils.js:467]: > I was thinking it could be a module-scope function. Then only that function could be exported and the actual CryptoKeyStore class could be entirely private within the module (not exported). So the only way to access the CryptoKeyStore would be by calling `newCryptoKeyStore()` and there would be no possible way (without subverting the Node module system) for any calling code to fail to initialize it correctly.
> 
> That does assume that no other code needs to directly access the CryptoKeyStore class though, such as to extend it. I am not sure whether or not that is the case.

Gotcha, but I prefer to keeping the class exported, based on idea that it is developer's responsibility to handle `initialize` action (when and how), while we also provide a method to get state-ready instance via utils.js.

**Comment by bestbeforetoday** [comment]
I think a recently merged pull request by @harrisob added an HSM test to the ts-scenario tests. Since you've had to modify the old HSM test, this change may well require some changes to the current ts-scenario test code too. It's probably worth rebasing on the latest master and rebuilding to be sure.

**Comment by davidkhala** [comment]
> I think a recently merged pull request by @harrisob added an HSM test to the ts-scenario tests. Since you've had to modify the old HSM test, this change may well require some changes to the current ts-scenario test code too. It's probably worth rebasing on the latest master and rebuilding to be sure.

Noted. WIP

**Comment by davidkhala** [comment]
It looks no obvious conflict during rebase. Let's wait for azure

File: fabric-ca-client/lib/FabricCAServices.js
Comment (bestbeforetoday) [inline fabric-ca-client/lib/FabricCAServices.js:75]: I guess the key-value store and CryptoKeyStore are being created explicitly here rather than calling `utils.newCryptoKeyStore()` because it needs to be done synchronously within a constructor. This means that `await cks.initialize()` doesn't get called as it normally would, but this is OK because we know that all that call does is call `await kvs.initialize()` and initialize is a no-op on InMemoryKeyValueStore. That is a lot of information required to see that this code is correct, and not at all obvious. This makes me very nervous about maintainability!

**Review by bestbeforetoday — COMMENTED** [review]
After wrapping my head around this latest revision I am less sure that the behaviour of `Utils.newCryptoKeyStore()` is helpful. The reason it needs to be async is so it can await on the call to initialize() on the CryptoKeyStore, and this is only there to make an async call to initialize() on the key-value store implementation. Our default key-value store implementation (InMemoryKeyValueStore) doesn't require any initialization so it is only there for the case where the caller has passed in a constructor function for us to new up a custom key-value store instance.

Why do we need them to pass in a constructor function and have us call new and initialize()? Why can't they just pass in a key-value store instance? Then they can deal with instantiation and any required initialization, and `Utils.newCryptoKeyStore()` could be a synchronous function that just takes an optional key-value store instance parameter.

So I am wondering if we could simply have:
```
module.exports.newCryptoKeyStore = (keyValueStore = new InMemoryKeyValueStore()) => {
    return new CryptoKeyStore(keyValueStore);
}
```

What do you think?

**Comment by davidkhala** [comment]
> After wrapping my head around this latest revision I am less sure that the behaviour of `Utils.newCryptoKeyStore()` is helpful. The reason it needs to be async is so it can await on the call to initialize() on the CryptoKeyStore, and this is only there to make an async call to initialize() on the key-value store implementation. Our default key-value store implementation (InMemoryKeyValueStore) doesn't require any initialization so it is only there for the case where the caller has passed in a constructor function for us to new up a custom key-value store instance.
> 
> Why do we need them to pass in a constructor function and have us call new and initialize()? Why can't they just pass in a key-value store instance? Then they can deal with instantiation and any required initialization, and `Utils.newCryptoKeyStore()` could be a synchronous function that just takes an optional key-value store instance parameter.
> 
> So I am wondering if we could simply have:
> 
> ```
> module.exports.newCryptoKeyStore = (keyValueStore = new InMemoryKeyValueStore()) => {
>     return new CryptoKeyStore(keyValueStore);
> }
> ```
> 
> What do you think?

Brutal and I love it. It make more sense based on the fact: we now design to provide storing options more from fabric-network#Wallet side and fabric-common in a stateless code pattern

File: fabric-ca-client/lib/FabricCAServices.js
Comment (davidkhala) [reply fabric-ca-client/lib/FabricCAServices.js:75]: To make it easy and maintainable, I think we could even skipping this part. So that we does not allow undefined cryptoSuite as parameter, developer have to provide it explicitly meanwhile ensure the cryptoSuite is read to use (with an initialized cryptoKeystore) . We will never assign a default implementation to them

**Comment by davidkhala** [comment]
BTW, global config 'key-value-store' is meant to provide path of developer customized global implementation class. I hope I could get rid of it in 2.x.

**Comment by davidkhala** [comment]
@bestbeforetoday Here is a breaking update and ready for review. ;-D

File: fabric-ca-client/lib/FabricCAServices.js
Comment (bestbeforetoday) [inline fabric-ca-client/lib/FabricCAServices.js:46]: This JSDoc shows `cryptoSuite` as optional but the code (and tests) now make it mandatory. We need to be careful about making breaking changes to existing public APIs. Is it practical (and does it make sense) to keep this as optional with a default value if not specified?

File: fabric-ca-client/types/index.d.ts
Comment (bestbeforetoday) [inline fabric-ca-client/types/index.d.ts:8]: `cryptoSuite` is optional here too, but mandatory in the code. It's a bit strange having a mandatory property following several optional ones too

File: fabric-ca-client/types/index.d.ts
Comment (bestbeforetoday) [inline fabric-ca-client/types/index.d.ts:66]: Using the tab indent is good for consistency with the rest of the code. The following lines of the comment still look to have spaces as indentation though

File: fabric-common/lib/Utils.js
Comment (bestbeforetoday) [inline fabric-common/lib/Utils.js:457]: I guess it probably doesn't matter but do we need to do the require here or can it move to the top with the rest now?

File: fabric-common/lib/Utils.js
Comment (bestbeforetoday) [inline fabric-common/lib/Utils.js:66]: I think this can be of type `CryptoSetting`

File: fabric-common/types/index.d.ts
Comment (bestbeforetoday) [inline fabric-common/types/index.d.ts:14]: I think the options can be of type `CryptoSetting`

File: test/integration/fabric-ca-certificate-service-tests.js
Comment (bestbeforetoday) [inline test/integration/fabric-ca-certificate-service-tests.js:32]: Same as a previous comment but the integration tests should generally be what an application developer might write and it worries me that the application code has to change to avoid failure when migrating to a version of the SDK that includes this change

**Review by bestbeforetoday — CHANGES_REQUESTED** [review]
Comments are mostly just very minor tweaks or questions but I'm worried that this change as-is may break existing application code by making a previously optional parameter mandatory

File: fabric-ca-client/lib/FabricCAServices.js
Comment (davidkhala) [reply fabric-ca-client/lib/FabricCAServices.js:46]: And further more,  is it good to accept two fashion of parameter input? people could provide either a string or an object to 1st parameter, And the handling process diverge from its type

I think we should only accept single fashion. Then both document and type definition could be clear.

File: fabric-common/lib/Utils.js
Comment (davidkhala) [reply fabric-common/lib/Utils.js:457]: Accept

**Comment by davidkhala** [comment]
@bestbeforetoday 
I guess I did some mistake. 

And we could get a clear view if we stick on assumption that: 
**in fabric-common, both us and people should only use or accept InMemoryCKS**

1. `cks.initialize()`, it could be deprecated. It was prepared for CouchDB or file-based CKS to avoid having promise inside CKS constructor

2. We could provide a default cryptoSuite with InMemoryCKS in ca-client. I could revert it back.

File: fabric-ca-client/types/index.d.ts
Comment (davidkhala) [reply fabric-ca-client/types/index.d.ts:66]: I lint this ts in another PR #239 
So we could hold this before 239 is merged

Comment (davidkhala) [reply fabric-ca-client/types/index.d.ts:66]: I apply tsLint to it in another PR #239 
So hold this before 239 is merged

File: fabric-ca-client/types/index.d.ts
Comment (bestbeforetoday) [inline fabric-ca-client/types/index.d.ts:11]: I'm not a fan of spacing everything out with every second line a blank line, but maybe that's just me :)

**Review by bestbeforetoday — APPROVED** [review]
👍

File: fabric-ca-client/types/index.d.ts
Comment (davidkhala) [reply fabric-ca-client/types/index.d.ts:11]: Aha, my editor could be modified to align with your preference. 
(It is Webstorm default config)

**Comment by btl5037** [comment]
Took us a bit to figure out what broke our performance tools, but I'll note, this is a breaking change, removing the synchronicity of the `newDefaultKeyValueStore()` made all of our code start failing as we received the promise returned by the call to `newDefaultKeyValueStore()` which now no longer occurs.

Just an FYI that this will need to go into the `2.2.0` release notes.

**Comment by bestbeforetoday** [comment]
I think we probably need to return `BaseClient.newDefaultKeyValueStore()` to being `async` to avoid a breaking change

**Comment by davidkhala** [comment]
> `ore()` made all of our

I guess apply `await` to all function could wipe out the uncertainty of whether a function is async or not. 
such as ` await Client.newDefaultKeyValueStore({path: testUtil.storePathForOrg(orgName)});`

So code on your side has a restriction that it have to be an async ?

**Pull request merged**
